hudi-common/src/main/java/org/apache/hudi/common/model/HoodieFileGroup.java

```diff
@@ -125,8 +133,21 @@ private boolean isFileSliceCommitted(FileSlice slice) {
     if (!compareTimestamps(slice.getBaseInstantTime(), LESSER_THAN_OR_EQUALS, lastInstant.get().getTimestamp())) {
       return false;
     }
+    // if its part of completed timeline, return true.
+    if (timeline.containsInstant(slice.getBaseInstantTime())) {
+      return true;
+    }
 
-    return timeline.containsOrBeforeTimelineStarts(slice.getBaseInstantTime());
+    if (firstActiveInstant.isPresent() && compareTimestamps(slice.getBaseInstantTime(), GREATER_THAN_OR_EQUALS, firstActiveInstant.get().getTimestamp())) {
+      // To get here:
+      // 1. the timestamp must be <= the last commit
+      // 2. not in the completed timeline
+      // 3. the timestamp must be >= the first active instant
```

Comment:
Yeah, the `#isBeforeTimelineStarts` is only valid when the timeline is the whole complete active timeline.